src/raftstore/store/worker/apply.rs

```diff
@@ -1710,6 +1698,64 @@ impl ApplyDelegate {
         ssts.push(sst.clone());
         Ok(Response::new())
     }
+
+    fn convert_range(&self, s_key: &[u8], e_key: &[u8]) -> Result<(Vec<u8>, Vec<u8>)> {
+        // start key must less than end key, the maximum end key is empty.
+        if !e_key.is_empty() && s_key >= e_key {
+            return Err(box_err!(
+                "invalid start_key: {:?}, end_key: {:?}",
+                s_key,
+                e_key
+            ));
+        }
+
+        // check start key
+        check_data_key(s_key, &self.region)?;
+        let start_key = keys::data_key(s_key);
+
+        // check end key
+        let end_key = keys::data_end_key(e_key);
+        let region_end_key = keys::data_end_key(self.region.get_end_key());
+        if end_key > region_end_key {
+            return Err(Error::KeyNotInRegion(e_key.to_vec(), self.region.clone()));
+        }
+
+        Ok((start_key, end_key))
+    }
+
+    fn convert_cf(cf: &str) -> Result<&str> {
+        if cf.is_empty() {
+            Ok(CF_DEFAULT)
+        } else if ALL_CFS.iter().find(|x| **x == cf).is_none() {
+            Err(box_err!("invalid unsafe cleanup command, cf: {:?}", cf))
+        } else {
+            Ok(cf)
+        }
+    }
+
+    fn handle_unsafe_cleanup_range(&mut self, req: &Request) -> Result<Response> {
+        let cf = Self::convert_cf(req.get_unsafe_cleanup_range().get_cf())?;
+        let (start_key, end_key) = self.convert_range(
+            req.get_unsafe_cleanup_range().get_start_key(),
+            req.get_unsafe_cleanup_range().get_end_key(),
+        )?;
+
+        // Unsafe_cleanup_range tasks are expensive operations, so we deliver them to
+        // another worker thread to prevent them blocking apply thread.
```

Comment:
This is highly unsafe and seems no defense check is added. How to make sure data are consistent after this command is executed?

Reply:
Any suggestion?

Reply:
Seems this can fail consistency check.

Reply:
if we can ensure the consistency check is run after all peers finishing the unsafe cleanup, it may not be a problem.

Reply:
I think we can set this range as `deprecated` state, and all writes on this range will fail, all consistency check on this range will return 0.